### PR TITLE
feat: support running k8s autodiscover suite for Beats PRs and local repositories

### LINKED ISSUE #1106: [k8s Autodiscover] SUpport running test suite agains Beats PRs or local filesystem
The same way we do for other test suites (fleet and metricbeat) we'd like to test beats' PRs/branches or binaries created from the local build in Beats.

For that reason, we'd need to detect if the build was triggered by a PR/branch from Beats, or using local file system, and load the Docker image for:
- metricbeat
- filebeat
- heartbeat
in Kind's runtime, using `kind load image-archive $theFile.tar`

cc/ @jsoriano

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
It adds same logic as in other suites to detect if we are working with a build triggered by Beats (a merge commit, or a PR) or if a developer is using the BEATS_LOCAL_PATH to run the tests for the locally built artefacts.

With that in mind, if we are using 1) CI snapshots or 2) testing local artifacts, we will 1) download the TAR file representing the Docker image for a given commit, or 2) use the local filesystem to get an URI to the local file.

Once we have the TAR.GZ file, we will load into the current Docker host, tagging the image, and then loading into kind using `kind load docker-image` command. This is needed because the k8s node provided by kind does not know anything about the images loaded into the Docker host. Thankfully, kind is usable enough to support loading images in this manner.

Finally, we are avoiding the download of the binary from the bucket with every scenario, but instead caching the beat version for a beat. This way we will skip the download&tag&load process, reducing the build time and the amount of bytes downloaded from GCP Storage.

## Why is it important?
This PR will allow Beats PRs and merges to run this test suite. Besides that, Beats developers will be able to run the e2e tests for k8s autodiscover from their local machines against their local binaries (Docker images produced by Beats' build system).

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have run the Unit tests for the CLI, and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## How to test this PR locally

**Running the tests for the `f1fea95e8d44b5b6c45a7d25026e6276f7248456` commit:**

```shell
$ SUITE="kubernetes-autodiscover" TAGS="filebeat" TIMEOUT_FACTOR=3 BEATS_USE_CI_SNAPSHOTS=true GITHUB_CHECK_SHA1=f1fea95e8d44b5b6c45a7d25026e6276f7248456 LOG_LEVEL=TRACE BEATS_USE_CI_SNAPSHOTS=true ELASTIC_APM_ACTIVE=false DEVELOPER_MODE=true make -C e2e functional-test
```

**Running the tests for a version:**
```shell
SUITE="kubernetes-autodiscover" TAGS="filebeat" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE ELASTIC_APM_ACTIVE=false DEVELOPER_MODE=true make -C e2e functional-test
```

## Related issues

- Closes #1106
- Completes #1064
- Completes https://github.com/elastic/beats/issues/25307

## Follow-ups
@adam-stokes we should simplify the usage of env vars: I can imagine removing `BEATS_USE_CI_SNAPSHOTS`: if `GITHUB_CHECK_SHA1` exists, we are using CI snapshots :)

### CODE REVIEW

File: e2e/_suites/kubernetes-autodiscover/autodiscover_test.go [hunk 1]
```diff
@@ -18,14 +22,28 @@ import (
...
 	"github.com/elastic/e2e-testing/internal/utils"
 )
 
+var beatVersions = map[string]string{}
+
 const defaultBeatVersion = "8.0.0-SNAPSHOT"
-const defaultEventsWaitTimeout = 120 * time.Second
-const defaultDeployWaitTimeout = 120 * time.Second
+
+var defaultEventsWaitTimeout = 60 * time.Second
```
Comment (mdelapenya) [inline e2e/_suites/kubernetes-autodiscover/autodiscover_test.go:37]: Using test framework default timeout factor of 3 (5 on CI)

**Review by cachedout — APPROVED** [review]

File: e2e/_suites/kubernetes-autodiscover/autodiscover_test.go [hunk 2]
```diff
@@ -159,10 +159,11 @@ func (m *podsManager) configureDockerImage(podName string) error {
 			return err
 		}
 
-		log.Tracef("Caching beat version '%s' for %s", beatVersion, podName)
-		beatVersions[podName] = beatVersion
 	}
 
+	log.Infof("Caching beat version '%s' for %s", beatVersion, podName)
```
Comment (mdelapenya) [inline e2e/_suites/kubernetes-autodiscover/autodiscover_test.go:164]:
```suggestion
	log.Tracef("Caching beat version '%s' for %s", beatVersion, podName)
```

**Review by jsoriano — COMMENTED** [review]
LGTM, thanks!

File: e2e/_suites/kubernetes-autodiscover/autodiscover_test.go [hunk 3]
```diff
@@ -18,14 +22,28 @@ import (
...
 const defaultBeatVersion = "8.0.0-SNAPSHOT"
-const defaultEventsWaitTimeout = 120 * time.Second
-const defaultDeployWaitTimeout = 120 * time.Second
+
+var defaultEventsWaitTimeout = 60 * time.Second
+var defaultDeployWaitTimeout = 60 * time.Second
+
+func init() {
+	// initialise timeout factor
+	common.TimeoutFactor = shell.GetEnvInteger("TIMEOUT_FACTOR", common.TimeoutFactor)
```
[the remark this answers is not in the document]
Comment (mdelapenya) [reply e2e/_suites/kubernetes-autodiscover/autodiscover_test.go:42]: Mmm, but then it's not ensured that the init method will be called in the proper order, right? IIRC Golang is not deterministic when talking about init functions order

[the remark this answers is not in the document]
Comment (mdelapenya) [reply e2e/_suites/kubernetes-autodiscover/autodiscover_test.go:42]: If we still want to do so, I'd move this file's init code to the beforeSuite phase.

[the remark this answers is not in the document]
Comment (jsoriano) [reply e2e/_suites/kubernetes-autodiscover/autodiscover_test.go:42]: > Mmm, but then it's not ensured that the init method will be called in the proper order, right? IIRC Golang is not deterministic when talking about init functions order

There are some guarantees in order, for example all the inits in an imported package are executed before continuing, or if a file has multiple inits, they are executed in order of definition.
For this case, if `init()` is defined in `common`, any package importing `common` will have this `init()` executed before running its own code.

**Pull request merged**